lib/charms/opensearch/v0/opensearch_plugins.py

```diff
@@ -0,0 +1,277 @@
+    def needs_upgrade(self) -> bool:
+        """Returns if the plugin needs an upgrade or not.
+
+        Needs upgrade must be set if an upgrade on the charm happens and plugins must
+        be updated.
+
+        Consider overriding this method only if the plugin needs an special care at upgrade.
+        """
+        current_version = self.version
+        # Now check if the format of this plugin and OpenSearch's match
+        num_points = len(self.distro.version.split("."))
+        return self.distro.version != current_version[:num_points]
+
+    @abstractmethod
+    def upgrade(self, uri: str) -> None:
+        """Runs the upgrade process in this plugin."""
+        pass
+
+    @abstractmethod
+    def is_enabled(self) -> bool:
+        """Returns True if the plugin is enabled."""
+        pass
+
+    @abstractmethod
+    def disable(self) -> bool:
+        """Disables the plugin.
+
+        Runs the configure() method with the configuration to disable the plugin.
+        """
+        pass
+
+    @abstractmethod
+    def enable(self) -> bool:
+        """Enables the plugin.
+
+        Runs the configure() method with the configuration to enable the plugin.
+        """
+        pass
+
+    def configure(
+        self,
+        opensearch_yml: Dict[str, str],
+        keystore: Dict[str, str] = {},
+    ) -> bool:
+        """Sets the plugin configuration. Returns True if restart needed."""
+        if not self._is_started():
+            return False
+        self._update_keystore_and_reload(keystore)
+        return any(
+            [
+                self.distro.config.put(self.CONFIG_YML, c, v)[c] == v
+                for c, v in opensearch_yml.items()
+            ]
+        )
+
+    def is_installed(self) -> bool:
+        """Returns True if the plugin name is present in the list of installed plugins."""
+        if not self._is_started():
+            return False
+        return self.name in self.distro.list_plugins()
+
+    def install(self, uri: str, batch=True) -> bool:
+        """Installs the plugin specified in the URI.
+
+        The URI can have one of the following formats:
+        * A zip file
+        * An URL that downloads a zip file
+        * A maven coded dependency
+        """
+        if self.is_installed():
+            # It is already available as a plugin, return True
+            return True
+        if self._depends_on:
+            plugins = self.distro.list_plugins()
+            missing = []
+            for dependency in self.depends_on:
+                if dependency not in plugins:
+                    missing.append(dependency)
+                if dependency:
+                    raise OpenSearchPluginError("Missing dependencies")
+        self.distro.add_plugin_without_restart(uri, batch=batch)
+
+    @property
+    @abstractmethod
+    def depends_on(self) -> List[str]:
+        """Returns a list of plugins it depends on."""
+        pass
+
+    def get_status(self) -> Union[int, StatusBase]:
```

Comment:
Not a fan of using `StatusBase` - it is a UI related class of the charm, this shouldn't be used for internal states of plugins etc.. We should create an enum for the plugins' states:
```
class PluginState(BaseStrEnum):
    MISSING = "missing"
    DISABLED = "disabled"
    WAITING_FOR_UPGRADE = "waiting-for-upgrade"
    ACTIVE = "active"


@property
def state() -> PluginState:
    if not self.installed():
        status = PluginState.MISSING
    elif not self.enabled():
        status = PluginState.DISABLED
    elif self.needs_upgrade():
        status = PluginState.WAITING_FOR_UPGRADE
    else:
        status = PluginState.ACTIVE

    return status
```